src/biz/collab/queue.rs

```diff
@@ -0,0 +1,467 @@
+  remove_pending_meta, storage_cache_key,
+};
+
+use crate::state::RedisConnectionManager;
+use anyhow::{anyhow, Context};
+use app_error::AppError;
+use collab_entity::CollabType;
+use database_entity::dto::{CollabParams, QueryCollab, QueryCollabResult};
+use serde::{Deserialize, Serialize};
+
+use std::cmp::Ordering;
+use std::collections::{BinaryHeap, HashMap};
+use std::ops::DerefMut;
+
+use sqlx::PgPool;
+use std::sync::atomic::AtomicI64;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::Mutex;
+use tokio::time::{interval, sleep, sleep_until, Instant};
+use tracing::{error, info, trace, warn};
+
+/// The maximum size of a batch payload in bytes.
+const MAXIMUM_BATCH_PAYLOAD_SIZE: usize = 2 * 1024 * 1024;
+type PendingWriteHeap = Arc<Mutex<BinaryHeap<PendingWrite>>>;
+#[derive(Clone)]
+pub struct StorageQueue {
+  collab_cache: CollabCache,
+  connection_manager: RedisConnectionManager,
+  pending_write: PendingWriteHeap,
+  pending_id_counter: Arc<AtomicI64>,
+}
+
+impl StorageQueue {
+  pub fn new(collab_cache: CollabCache, connection_manager: RedisConnectionManager) -> Self {
+    let next_duration = Arc::new(Mutex::new(Duration::from_secs(2)));
+    let pending_id_counter = Arc::new(AtomicI64::new(0));
+    let pending_write = Arc::new(Mutex::new(BinaryHeap::new()));
+
+    // Spawns a task to recover pending write operations from the Redis cache. This task is crucial
+    // for system resilience: if the server restarts unexpectedly, pending writes stored in the cache
+    // are at risk of being lost. By fetching and repopulating these writes into the pending write heap,
+    // we ensure that no data is lost and that write operations can resume seamlessly post-restart.
+    spawn_fetch_remaining_pending_write(
+      connection_manager.clone(),
+      pending_write.clone(),
+      pending_id_counter.clone(),
+    );
+
+    // Spawns a task that periodically writes pending collaboration objects to the database.
+    spawn_period_write(
+      next_duration.clone(),
+      collab_cache.clone(),
+      connection_manager.clone(),
+      pending_write.clone(),
+    );
+
+    spawn_period_check_pg_conn_count(collab_cache.pg_pool().clone(), next_duration);
+
+    Self {
+      collab_cache,
+      connection_manager,
+      pending_write,
+      pending_id_counter,
+    }
+  }
+
+  /// Enqueues a object for deferred processing. High priority writes are processed before low priority writes.
+  ///
+  /// adds a write task to a pending queue, which is periodically flushed by another task that batches
+  /// and writes the queued collaboration objects to a PostgreSQL database.
+  ///
+  /// This data is stored temporarily in the `collab_cache` and is intended for later persistent storage
+  /// in the database. It can also be retrieved during subsequent calls in the [CollabStorageImpl::get_encode_collab]
+  /// to enhance performance and reduce database reads.
+  ///
+  pub async fn push(
```

Comment:
if possible, should check current load and flush to postgres if exceeds certain limit, this will mitigate scenario with sudden spike.